### PR TITLE
Runtime annotations

### PR DESCRIPTION
**Pull request by sfbwalder**
This is a first cut at allowing the use of annotations to specify which tests to run at run-time as was considered in #17.  The code needs some clean-up, so consider this an initial proposal that is very open to everyone's input.

The gist is this:
1. Put an annotation (e.g. @smoke) on a [feature](https://github.com/curit/ember-cli-yadda/compare/master...StateFarmIns:runtime-annotations?expand=1#diff-8e78871836b9805f5211c35b5f7a2e67R1) or [scenario](https://github.com/curit/ember-cli-yadda/compare/master...StateFarmIns:runtime-annotations?expand=1#diff-1433769e502699dc7c66a7a23afe43ebR3) in the .feature file.
2. [Specify which annotations to run](https://github.com/curit/ember-cli-yadda/compare/master...StateFarmIns:runtime-annotations?expand=1#diff-69f16599a8e1e70e9e2cbef7f505b706R42) in config/environment.js in the appropriate environment section.  This could be made more flexible by passing in an arg on `ember t` or using environment variables to specify the annotations to run.
3. The [acceptance test runner calls](https://github.com/curit/ember-cli-yadda/compare/master...StateFarmIns:runtime-annotations?expand=1#diff-b7285f87d2f21b09853931ca7d62fd6eR4) a *user defined module* (e.g. [yadda-run-test-helper.js](https://github.com/curit/ember-cli-yadda/compare/master...StateFarmIns:runtime-annotations?expand=1#diff-cea075e881a9755d4ef21f2df506ba04R1)), passing it the annotations on the feature or scenario.  The user defined module has the logic to compare the annotations on the test to the requested annotations in the configuration and returns true/false.  If true, the acceptance test runner runs the test; if false it does not run the test.
4. The user defined module is [configurable in ember-cli-build.js](https://github.com/curit/ember-cli-yadda/compare/master...StateFarmIns:runtime-annotations?expand=1#diff-2fe2a7faee7de1e0d5d82e032498d84cR8).  This configuration is [used by the feature parser](https://github.com/curit/ember-cli-yadda/compare/master...StateFarmIns:runtime-annotations?expand=1#diff-b63b6e2a30ddb59a52272126f6ffec65R44) to import the module if it exists.
 * Using a user defined module keeps the test runner from enforcing rules on how annotations are used.  

Please provide opinions and I will make updates to the code and this pull request as needed.

### CODE REVIEW

**Comment by albertjan** [comment]
Very cool! Thanks so much for this. I'm not entirely clear on how this would work for the user, but that will get clearer with more study-ing. I have one question now:

Do I understand correctly that if people want to handle custom annotations they'll need to create a `yadda-run-test-helper.js` file? If so it might be nice to add this file to the default blueprint, so it's created with the `skip` annotation by default so people can see how it works.

**Comment by sfbwalder** [comment]
> Do I understand correctly that if people want to handle custom annotations they'll need to create a yadda-run-test-helper.js file?

Yes.  They can name it what they want and put what ever logic in it they want.  If we want to provide a default implementation we could provide that as a blueprint.

**Comment by albertjan** [comment]
I think it would make sense to handle all annotations the same way.

**Comment by sfbwalder** [comment]
I simplified some things:
1. Annotations are all (including @ignore) handled in [tests/helpers/yadda-annotations.js](https://github.com/curit/ember-cli-yadda/pull/52/files#diff-cef3731c5847346502cc0cb24e828ab5R1).
    * The configuration in ember-cli-build.js has been removed.  The [feature parser](https://github.com/curit/ember-cli-yadda/pull/52/files#diff-b63b6e2a30ddb59a52272126f6ffec65R44) always imports tests/helpers/yadda-annotations.
    *  See [acceptance.js](https://github.com/curit/ember-cli-yadda/pull/52/files#diff-b7285f87d2f21b09853931ca7d62fd6eR3) for how it is called.
1. A default implementation of yadda-annotations.js is in the blueprint and will be put in the consuming app's tests/helpers at install time.
1. The default implementation provides backwards compatibility.  See comments in yadda-annotations.js for more details.
1. If the consumer needs something other than than the default implementation, they can change it in tests/helpers/yadda-annotations.js.

I didn't know how to change the ember's ENV.annotations between running feature acceptance tests. 
 Currently [config/environment.js](https://github.com/curit/ember-cli-yadda/pull/52/files#diff-69f16599a8e1e70e9e2cbef7f505b706R41) needs to be changed manually and the tests rerun to test all annotation scenarios.  See comments in config/environment.js for more info.  Thoughts on how to automate this?

Does any of this break mocha consumers?

Let me know your thoughts on where this is going, and I will continue to adjust and clean it up.

File: lib/test-runner/qunit/acceptance.js [hunk 1]
```diff
@@ -1,14 +1,12 @@
 function testFeature(feature) {
 
-  // if a runFeature function is provided, use it to determine if should run the feature
-  if (typeof(runFeature) === 'function' && !runFeature(feature.annotations)) {
-    console.log('not running feature: ', feature.title);
+  let runFeature = yaddaAnnotations.runFeature(feature.annotations);
```
Comment (albertjan) [inline lib/test-runner/qunit/acceptance.js]: if `runFeature` would return a function and if we'd call that function from the current scope would we be able to access all the stuff in the current scope?

i.e.:

```
if (annotations.ignore) {
  return function() {
    skip(`Feature ${feature.title}`, function (assert) {});
  }
}
```

This way you could do anything to the test, not just skip it. I'm not sure if we could teach this to the user though. We definitely should rely on them having to do this stuff: 

          test('Scenario: ${scenario.title}', function(assert) {
            let self = this;
            return new Ember.RSVP.Promise(function(resolve) {
              yadda.Yadda(library.default(assert), self).yadda(scenario.steps, { ctx: {} }, function next(err, result) {
                if (err) {
                  throw err;
                }
                resolve(result);
              });
            });

I don't know how that would work exactly. We will definitely have to change the mocha runner to do this annotations dance too though.

**Comment by sfbwalder** [comment]
Is this anywhere close to the concept you are thinking?
* [code here returns a function](https://github.com/curit/ember-cli-yadda/pull/52/files#diff-396322b0f44cb1c286e305423d02b02eR12)
* [code here executes it](https://github.com/curit/ember-cli-yadda/pull/52/files#diff-b7285f87d2f21b09853931ca7d62fd6eR3)

With the way it is coded, there would need to be a decision on what params to pass to the functions.  Thoughts?

**Comment by sfbwalder** [comment]
Thoughts on where this is going?  I'll clean it up and add some doc if so.

**Comment by albertjan** [comment]
Yeah definitely, I'll try to have a look this evening. I think this is a very worth while extension. We'll have to look into mocha support and upgrade path, from the top of my head. I'll look into this asap. 😄

**Comment by albertjan** [comment]
I looked at it yesterday and got kind of lost in getting ember try to install `ember-cli-mocha` to test the mocha stuff and then my npm / node install disintegrated. 👍  

I really like what you've done here and I think it's the correct way to handle annotations. Giving sensible defaults and providing extensibility for future use.

@simonihmig and @adamjmcgrath and @mschinis If you guys have time and are still using ember-cli-yadda do you want to read through this and see what you think 😄

**Comment by adamjmcgrath** [comment]
This branch is failing for mocha because of #53 if you rebase it against master I'll take a look

**Comment by mschinis** [comment]
hey @albertjan, we're currently not using ember-cli-yadda 😊

**Comment by sfbwalder** [comment]
I'm not sure I re-based this correctly :slightly_frowning_face:.

@adamjmcgrath can you try this out for mocha?

**Comment by adamjmcgrath** [comment]
```bash
$ npm install -g ember-cli@2.16.2
$ ember new yaddatest
$ cd yaddatest/
$ ember install ember-cli-mocha
$ ember install StateFarmIns/ember-cli-yadda#runtime-annotations
$ ember g ember-cli-yadda
$ ember g feature make-feature
$ ember t
...
not ok 1 Chrome 63.0 - TestLoader Failures yaddatest/tests/acceptance/make-feature-test: could not be loaded
    ---
        message: >
            Could not find module `yaddatest/tests/helpers/yadda-annotations` imported from `yaddatest/tests/acceptance/make-feature-test`
```

I guess you need a mocha equivalent to [blueprints/qunit/ember-cli-yadda/files/tests/helpers/yadda-annotations.js](https://github.com/StateFarmIns/ember-cli-yadda/blob/87b45f53e1289d8b5e40f4c9b6e7cdb0265a0360/blueprints/qunit/ember-cli-yadda/files/tests/helpers/yadda-annotations.js)

**Comment by sfbwalder** [comment]
@adamjmcgrath I added `yadda-annotations` for mocha, then followed the steps you did and it worked.  I then did some additional manual testing to
-  ensure it was backwards compatible with @ignore.
-  included / excluded features and scenarios for other annotations

It seem to be working.  Can you please try again?

**Comment by adamjmcgrath** [comment]
```
tests/helpers/yadda-annotations.js
  47:3  error  Unexpected console statement  no-console
```

Other than that, LGTM

**Comment by sfbwalder** [comment]
Tests are failing when attempting to run with ember 3.0.  I think this add-on needs updated to 2.x.  Should I open an issue and do a PR?

**Comment by albertjan** [comment]
@eltjo-k Hey man! How are you doing? Can you give these guys write access to this repo? I don't have the time to maintain it at the moment.

**Comment by albertjan** [comment]
Alright guys I made you both contributors. if you can decide amongst yourselves that it works merge at will :+1: let me know when I should do a npm release. 😄

**Comment by sfbwalder** [comment]
@albertjan @adamjmcgrath I think this is ready to merge.  Any objections?

**Comment by albertjan** [comment]
Go for it.

________________________________

**Comment by sfbwalder** [comment]
@albertjan ready for new release, thx

**Comment by albertjan** [comment]
@sfbwalder do you have a npm account?

**Pull request merged**